src/bot/commands/ping-pinger/model.js

```diff
@@ -0,0 +1,17 @@
+const mongoose = require("mongoose");
+
+// See: https://stackoverflow.com/questions/19051041/cannot-overwrite-model-once-compiled-mongoose
+
+let model;
+
+try {
+  model = mongoose.model("Pinger");
+} catch (error) {
```

Comment:
Can you explain why this try/catch is necessary? I see in that stackoverflow question it has to do with not being able to redefine the model, but I thought Node only executes files the first time they're required and caches their exports, so I'm a bit confused about why this is necessary here

Reply:
I actually thought the exact same thing at first. This is what I initially tried:

```js
const mongoose = require("mongoose");

module.exports = mongoose.model("Pinger", {
  guildId: String,
  id: String,
  url: String,
});
```

Which gave me this error:

```
/home/user/nce_discord_bot/node_modules/mongoose/lib/index.js:543
      throw new _mongoose.Error.OverwriteModelError(name);
      ^

OverwriteModelError: Cannot overwrite `Pinger` model once compiled.
    at Mongoose.model (/home/user/nce_discord_bot/node_modules/mongoose/lib/index.js:543:13)
    at Object.<anonymous> (/home/user/nce_discord_bot/src/bot/commands/ping-pinger/model.js:5:27)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/home/user/nce_discord_bot/src/bot/commands/ping-pinger/ping.js:2:16)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at CommandHandler.load (/home/user/nce_discord_bot/node_modules/discord-akairo/src/struct/AkairoHandler.js:128:26)
```

Originally I didn't really care about the issue because it was a relatively easy "fix", but since you mentioned it, I couldn't stop thinking about why it wasn't using a cached version. So I decided to do a bit of debugging and find out why.

First I tried to diff the keys of `require.cache` to see if it's following the [caching requirements](https://nodejs.org/api/modules.html#modules_caching) (tl;dr: if the modules resolve to the same path and the cache hasn't been modified, it should use the cached version). No dice.

I then thought about how Akairo's handler class, [`AkairoHandler`](https://discord-akairo.github.io/#/docs/main/master/class/AkairoHandler), dynamically `require`s our command modules ([`CommandHandler`](https://discord-akairo.github.io/#/docs/main/master/class/CommandHandler), which we're using, inherits the functionality). I ran our bot through a debugger and realized Akairo was loading `./src/bot/commands/ping-pinger/model.js` because [we're loading all `*.js` files in `./src/bot/commands/` as Akairo modules/commands apparently](https://github.com/nationalcollegiateesports/discord-bot/blob/285c7359ca43f9b16c986fe07a85256ae3fd8da8/bin/bot.js#L19). Upon realizing that it's not a class inheriting from [`AkairoModule`](https://discord-akairo.github.io/#/docs/main/master/class/AkairoModule) (e.g. [`Command`](https://discord-akairo.github.io/#/docs/main/master/class/Command)), it [purges it from the cache](https://github.com/discord-akairo/discord-akairo/blob/29810b9e5d5cd804c649948ad60e5f2ac3897141/src/struct/AkairoHandler.js#L133).

Mystery solved. Now what? I think we just need to decide if we want to stick with the `try`/`catch` solution or move the models somewhere else in our project. Alternatively, there might be a way to fine-tune what exactly Akairo loads, but I haven't checked. I don't really have a preference what we do though. Thoughts?

Reply:
Wack. I've definitely heard of command frameworks manually purging the cache to load files though (I think my Eris-based framework does something similar, actually), so I guess that does make some sense.

I suppose the other way to solve this would be to move the model definition out of the commands folder - this would mean that we wouldn't be able to keep the DB model and the command using it as close to each other, so we'd need to compare trade-offs between that organization style and the alternative where we just have all the models in a `models` directory on their own. Being able to do away with this weirdness would be a plus, and I'm not particularly attached to the current directory structure, but I think it should be your call.

Reply:
I personally wasn't familiar with the practice but yeah, it does make sense, and it also explains how Akairo is able to hot-reload modules. Interesting.

I'm fine with moving it to `./src/bot/models/`. It's still an obvious location, it removes the weirdness, and it could make it feel less awkward to use the same models/collections across multiple commands.